src/display/editor/draw.js

```diff
@@ -0,0 +1,832 @@
+/* Copyright 2022 Mozilla Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { AnnotationEditorParamsType, unreachable } from "../../shared/util.js";
+import { AnnotationEditor } from "./editor.js";
+import { noContextMenu } from "../display_utils.js";
+
+class DrawingOptions {
+  #svgProperties = Object.create(null);
+
+  updateProperty(name, value) {
+    this[name] = value;
+    this.updateSVGProperty(name, value);
+  }
+
+  updateProperties(properties) {
+    if (!properties) {
+      return;
+    }
+    for (const [name, value] of Object.entries(properties)) {
+      this.updateProperty(name, value);
+    }
+  }
+
+  updateSVGProperty(name, value) {
+    this.#svgProperties[name] = value;
+  }
+
+  toSVGProperties() {
+    const root = this.#svgProperties;
+    this.#svgProperties = Object.create(null);
+    if (!root) {
+      this.updateProperties(this);
+      return this.toSVGProperties();
+    }
+    return { root };
+  }
+
+  reset() {
+    this.#svgProperties = null;
+  }
+
+  clone() {
+    unreachable("Not implemented");
+  }
+}
+
+/**
+ * Basic draw editor.
+ */
+class DrawingEditor extends AnnotationEditor {
+  #drawOutlines = null;
+
+  #mustBeCommitted;
+
+  static _currentDrawId = -1;
+
+  static _currentDraw = null;
+
+  static _currentDrawingOptions = null;
+
+  static _INNER_MARGIN = 3;
+
+  constructor(params) {
+    super(params);
+    this._drawId = null;
+    this.#mustBeCommitted = params.mustBeCommitted || false;
+
+    if (params.drawOutlines) {
+      this.#createDrawOutlines(params);
+      this.#addToDrawLayer();
+    }
+  }
+
+  #createDrawOutlines({ drawOutlines, drawId, drawingOptions }) {
+    this.#drawOutlines = drawOutlines;
+    this._drawingOptions ||= drawingOptions;
+
+    if (drawId >= 0) {
+      this._drawId = drawId;
+      // We need to redraw the highlight because we change the coordinates to be
```

Comment:
Should this comment still have "highlight" here, now that this is a more general purpose class?